### PR TITLE
Handle large snapshots 

### LINKED ISSUE #1858: Snapshots may be larger than ring buffer maximum message size
As a CCF service processes more transactions which write to different tables at different keys, snapshots may grow without bounds. Currently, when a snapshot is generated, it is written to the outbound ring buffer in one go. It might be split into fragments if it is too large but if it actually is larger than the maximum message size (`--max-msg-size` option to `cchost`), the CCF node will crash, e.g. (with `16777216` the default maximum message size):

```
2020-11-04 15:22:18.596 | ERROR    | infra.remote:log_errors:86 - Contents of /home/lyshi/repos/CCF/tests/workspace/recovery_0/err:
terminating with uncaught exception of type std::logic_error: Requested a write of 16827539 bytes, max allowed is 16777216
```

To solve this, we could just bump the default maximum message size but this just pushes the problem further. The solution to this should work for any snapshot size, however large it is. Instead, we could either:
- Split the snapshot into multiple chunks before writing it to the ring buffer.
- Build a enclave<->host large message system which lets the enclave directly write to host memory, bypassing the ring buffer altogether. This would involve allocating a sufficiently large buffer on the host to store the snapshot. It seems that this solution should only apply to snapshots as other ring buffer writers already chunk messages into smaller bits (e.g. TLS). 

**Notes:**
- This issue should not affect services with small key-value store footprint, e.g. services that write to a few tables and/or a few keys. 
- To reproduce this issue:
```bash
$ TEST_ENCLAVE=virtual python recovery.py --gov-script CCF/tests/gov.lua --use-snapshot --snapshot-tx-interval 1000 --msgs-per-recovery 1000000
```

### PR DESCRIPTION
**Pull request by jumaffre**
Resolves #1858 

The maximum allowed size of snapshots was previously limited by the maximum allowed size of messages over the ring buffer. To cater for applications with large KV footprints, this PR adds support for large snapshots by directly copying snapshots from enclave to host memory, rather than going over the ring buffer. 

Some design points:
- Because of the new enclave <-> host roundtrip to allocate a snapshot buffer in host memory, it is possible that a busy host takes a long time to return the address of the allocated buffer to the enclave (e.g. because of high TCP traffic). In such case, i) pending snapshots may pile up in the enclave or ii) the corresponding snapshot evidence may be committed before the enclave has actually stored the snapshot in host memory. For i), we limit the number of pending snapshots at a given time so that the enclave doesn't run out of memory (new snapshots are simply not generated). For ii), we make sure that only snapshots that have been stored on the host (`is_stored`) can be committed (a later commit will eventually flush the pending snapshot).
- **Un**commtted snapshot files are no longer written to disk. Instead, they are kept in host memory until the snapshot evidence is committed (typically a short period). Because the enclave only keeps up to 10 pending snapshots in memory, the host will not keep more than 10 snapshots in memory either (happy to discuss this number). This is to prevent having an extra enclave -> host ring buffer message for the enclave to signal that is has written the snapshot to the host-allocated memory. 

Sequence diagram:
```mermaid

sequenceDiagram

participant Disk
participant Host
participant Enclave

Enclave->>Enclave: Take store snapshot at N.
Enclave->>Enclave: Serialise snapshot and add to pending_snapshots (enclave)
Enclave->>Host: [NEW] Request buffer with size S = snapshot serialised size
Host->>Host: Allocate buffer of size S and add to pending_snapshots (host)
Host->>Enclave: [NEW] Pointer to host-allocated memory
Enclave-->>Enclave: On SGX, sanitise buffer
Enclave-->>Host: Writes serialised snapshot at pointer
Enclave->>Enclave: Snapshot evidence is committed
Enclave->>Host: Writes snapshot receipt on ring buffer
Host->>Disk: Write snapshot + receipt to disk
```

TODO: 
- [x] Handle out-of-order snapshot/commit
- [x] End-to-end test
- [x] Fix unit test

### CODE REVIEW

**Review by achamayou — APPROVED** [review]

File: tests/e2e_operations.py [hunk 1]
```diff
@@ -150,6 +150,55 @@ def test_forced_snapshot(network, args):
...
 
 
+# https://github.com/microsoft/CCF/issues/1858
+@reqs.description("Generate snapshot larger than ring buffer max message size")
+def test_large_snapshot(network, args):
+    primary, _ = network.find_primary()
+
+    # Submit some dummy transactions
+    entry_size = 1000  # Lower bound on serialised write set size
+    iterations = int(args.max_msg_size_bytes) // entry_size
```
Comment (eddyashton) [inline tests/e2e_operations.py:160]: How many iterations is this in practice? We could easily end up sending 100s of requests here, resulting in a pretty slow test. If so, we should find a more efficient way to build a large snapshot (write-multiplying endpoint).

File: src/node/snapshotter.h [hunk 2]
```diff
@@ -257,6 +275,63 @@ namespace ccf
...
+            pending_snapshot.serialised_snapshot.begin(),
+            pending_snapshot.serialised_snapshot.end(),
+            snapshot_buf.begin());
+          pending_snapshot.is_stored = true;
+
+          LOG_DEBUG_FMT(
+            "Successfully copied snapshot at seqno {} to host memory [{}]",
+            it->first,
+            pending_snapshot.serialised_snapshot.size());
+          return true;
```
[the remark this answers is not in the document]
Comment (jumaffre) [reply src/node/snapshotter.h]: When the snapshot evidence is committed and the receipt has been passed to the host (currently `:223`).

File: tests/e2e_operations.py (hunk 1 above)
Comment (jumaffre) [reply tests/e2e_operations.py:160]: This was about 100 iterations, which took about 4 seconds in practice. I've bumped `entry_size` to `10000`, which reduces it to 10 iterations, or about half a second.

File: src/node/snapshotter.h [hunk 3]
```diff
@@ -157,17 +163,29 @@ namespace ccf
         return;
       }
 
+      auto evidence_version = tx.commit_version();
+
       pending_snapshots[snapshot_version].commit_evidence = commit_evidence;
       pending_snapshots[snapshot_version].write_set_digest = ws_digest;
       pending_snapshots[snapshot_version].snapshot_digest = cd.value();
+      pending_snapshots[snapshot_version].evidence_version = evidence_version;
+      pending_snapshots[snapshot_version].generation_count = generation_count;
```
[the remark this answers is not in the document]
Comment (achamayou) [reply src/node/snapshotter.h:172]: @jumaffre I don't believe we snapshot uncommitted states, so I don't think that's possible. Is that right?

File: src/node/snapshotter.h [hunk 4]
```diff
@@ -257,6 +275,63 @@ namespace ccf
...
+    bool store_snapshot(
+      std::span<uint8_t> snapshot_buf, uint32_t generation_count)
+    {
+      std::lock_guard<ccf::pal::Mutex> guard(lock);
+
+      for (auto it = pending_snapshots.begin(); it != pending_snapshots.end();
+           it++)
+      {
+        auto& pending_snapshot = it->second;
+        if (generation_count == pending_snapshot.generation_count)
```
[the remark this answers is not in the document]
Comment (jumaffre) [reply src/node/snapshotter.h]: Good point. Done!

File: src/node/snapshotter.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (jumaffre) [reply src/node/snapshotter.h]: Yep, that's right. We only create snapshots for committed states (the snapshotter is driven by `commit()`). This was the case before this PR already, and hasn't changed here.

**Pull request merged**